src/Features/CSharp/Portable/MakeLocalFunctionStatic/MakeLocalFunctionStaticHelper.cs

```diff
@@ -0,0 +1,147 @@
+using Microsoft.CodeAnalysis.PooledObjects;
+using Microsoft.CodeAnalysis.Shared.Extensions;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.CSharp.MakeLocalFunctionStatic
+{
+    internal static class MakeLocalFunctionStaticHelper
+    {
+        public static bool IsStaticLocalFunctionSupported(SyntaxTree tree)
+            => tree.Options is CSharpParseOptions csharpOption && csharpOption.LanguageVersion >= LanguageVersion.CSharp8;
+
+        public static bool TryGetCaputuredSymbols(LocalFunctionStatementSyntax localFunction, SemanticModel semanticModel, out ImmutableArray<ISymbol> captures)
+        {
+            var dataFlow = semanticModel.AnalyzeDataFlow(localFunction);
+            captures = dataFlow.CapturedInside;
+
+            return dataFlow.Succeeded;
+        }
+
+        public static async Task<Document> MakeLocalFunctionStaticAsync(
+            LocalFunctionStatementSyntax localFunction,
+            ImmutableArray<ISymbol> captures,
+            SemanticModel semanticModel,
+            Document document,
+            CancellationToken cancellationToken)
+        {
+            var root = (await document.GetSyntaxRootAsync(cancellationToken).ConfigureAwait(false))!;
+
+            var localFunctionSymbol = semanticModel.GetDeclaredSymbol(localFunction, cancellationToken);
+            var parameters = CreateParameterSymbols(captures);
+
+            // Finds all the call sites of the local function
+            var referencedSymbols = await SymbolFinder.FindReferencesAsync(
+                localFunctionSymbol, document.Project.Solution, cancellationToken).ConfigureAwait(false);
+
+            // Now we need to find all the refereces to the local function that we might need to fix.
+            var shouldWarn = false;
+            var invocationReferences = new List<InvocationExpressionSyntax>();
+
+            foreach (var referencedSymbol in referencedSymbols)
+            {
+                foreach (var location in referencedSymbol.Locations)
+                {
+                    // Since this is a local function, all reference must be in the same tree.
+                    var referenceNode = root.FindNode(location.Location.SourceSpan);
+                    if (!(referenceNode is IdentifierNameSyntax identifierNode))
+                    {
+                        continue;
+                    }
+
+                    if (identifierNode.Parent is InvocationExpressionSyntax invocation)
+                    {
+                        invocationReferences.Add(invocation);
+                    }
+                    else
+                    {
+                        // We won't be able to fix non-invocation references, 
+                        // e.g. creating a delegate. 
+                        shouldWarn = true;
+                    }
+                }
+            }
+
+            var nodeToTrack = new List<SyntaxNode>(invocationReferences) { localFunction };
```

Comment:
>new List<SyntaxNode> [](start = 30, length = 20)

We can avoid this list allocation.

Reply:
You mean a separate call to `TrackNodes` for `localFunction`? That would still create an array for the params though.

Reply:
note: if we use SyntaxEditor here (which also falls out if you do fix-all using SyntaxEditorBasedFixAllProvider), then i don't believe there's any need to TrackNodes at all.

Reply:
Fixed with SyntaxEditor